storage/src/main/java/io/atomix/storage/journal/FileChannelJournalSegmentReader.java

```diff
@@ -93,19 +92,23 @@ public long getNextIndex() {
   }
 
   @Override
-  public void reset(long index) {
+  public void reset(final long index) {
     reset();
-    Position position = this.index.lookup(index - 1);
-    if (position != null) {
+
+    final Position position = this.index.lookup(index - 1);
+    if (position != null && position.index() >= segment.index()) {
       currentEntry = new Indexed<>(position.index() - 1, null, 0);
       try {
         channel.position(position.position());
         memory.clear().flip();
       } catch (IOException e) {
         throw new StorageException(e);
       }
+
+      nextEntry = null;
```

Comment:
why we need to set this to null now?

Reply:
While not necessary in the happy path, if something goes wrong in `readNext`, then the state is weird - I'd rather not have something in `nextEntry` that does not make sense with the `currentEntry` we just set.

Reply:
hm ok, should we reset the current entry on the io exception?

Reply:
Good catch, yes!